### PR TITLE
Chore: Migrate LoginServices to Hooks

### PR DESCRIPTION
**Pull request by reinaldonetof**
## How to test or reproduce

- Join into `open.rocket.chat`
- Go to Login screen and test the oauth's buttons, `more options` and login with email too

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves a current function)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Documentation update (if none of the other choices apply)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat) doc
- [x] I have signed the [CLA](https://cla-assistant.io/RocketChat/Rocket.Chat.ReactNative)
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works (if applicable)
- [ ] I have added necessary documentation (if applicable)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/containers/LoginServices.tsx [hunk 1]
```diff
@@ -92,154 +94,145 @@ interface IOauthProvider {
...
+	const { length } = Object.values(services);
+
+	const heightButtons = useSharedValue(SERVICES_COLLAPSED_HEIGHT);
+
+	const animatedStyle = useAnimatedStyle(() => ({
+		overflow: 'hidden',
+		height: withTiming(heightButtons.value, { duration: 300, easing: Easing.inOut(Easing.quad) })
+	}));
+
+	const onPressFacebook = () => {
```
Comment (diegolmello) [inline app/containers/LoginServices.tsx:123]: On our refinement, we talked about extracting all these service logic functions away from the component.

File: app/containers/LoginServices.tsx [hunk 2]
```diff
@@ -271,77 +264,51 @@ class LoginServices extends React.PureComponent<ILoginServicesProps, ILoginServi
...
+			twitter: onPressTwitter,
+			wordpress: onPressWordpress
 		};
 		return oauthProviders[name];
 	};
 
-	renderServicesSeparator = () => {
-		const { collapsed } = this.state;
-		const { services, separator, theme } = this.props;
+	const renderServicesSeparator = () => {
```
Comment (diegolmello) [inline app/containers/LoginServices.tsx:285]: You don't want to use `renderSomething`.
That's ok for class components, but not for function components, as it's going to get recreated on every render.
The ideal fix is to extract the logic to its own component `<ServicesSeparator />` and send the necessary props.

A _not-so-valid_ fix would be to use `useCallback`, but it's hard to maintain and it's kind of anti-pattern.
There are use cases for this, but 99% of times a separate component is the right way.

File: app/containers/LoginServices.tsx [hunk 3]
```diff
@@ -271,77 +264,51 @@ class LoginServices extends React.PureComponent<ILoginServicesProps, ILoginServi
...
+						}}
 						style={styles.options}
 						color={themes[theme].actionTintColor}
 					/>
 					<OrSeparator theme={theme} />
 				</>
 			);
 		}
-		if (length > 0 && separator) {
+		if (length > 0 && Accounts_ShowFormLogin) {
```
Comment (diegolmello) [inline app/containers/LoginServices.tsx:305]: I didn't get this change.

**Review by diegolmello — CHANGES_REQUESTED** [review]

File: app/containers/LoginServices/index.tsx [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+		<>
+			{Object.values(services).map((service: IItemService) => (
+				<ServicesList
+					CAS_enabled={CAS_enabled}
+					CAS_login_url={CAS_login_url}
+					Gitlab_URL={Gitlab_URL}
+					server={server}
+					service={service}
+				/>
+			))}
```
Comment (diegolmello) [inline app/containers/LoginServices/index.tsx:73]: This is wrong.
`ServicesList` should be the whole map and render a `Service` component (called `ServiceList` now).

```tsx
// index.tsx
<ServicesList services={services} />

// ServicesList.tsx
{Object.values(services).map((service: IItemService) => (
	<Service
		CAS_enabled={CAS_enabled}
		CAS_login_url={CAS_login_url}
		Gitlab_URL={Gitlab_URL}
		server={server}
		service={service}
	/>
))}
```

File: app/containers/LoginServices/index.tsx [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+			<ServicesSeparator
+				services={services}
+				separator={separator}
+				collapsed={collapsed}
+				onPressButtonSeparator={onPressButtonSeparator}
```
Comment (diegolmello) [inline app/containers/LoginServices/index.tsx:78]: No need to name it using the context.

```suggestion
				onPress={onPressButtonSeparator}
```

File: app/containers/LoginServices/interfaces.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+	url: string;
+	ssoToken?: string;
+	authType?: string;
+}
+
+export interface IItemService {
+	_id: string;
+	name: string;
+	service: string;
+	authType: string;
```
Comment (diegolmello) [inline app/containers/LoginServices/interfaces.ts:11]: You know the available types.
We could benefit from having them defined here.

```suggestion
	authType: 'facebook' | 'github' | ...;
```

File: app/containers/LoginServices/serviceLogin.ts [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+import { Base64 } from 'js-base64';
+
+import { events, logEvent } from '../../utils/log';
+import { Services } from '../../lib/services';
+import Navigation from '../../lib/navigation/appNavigation';
+import { IItemService, IOpenOAuth, IServiceLogin } from './interfaces';
+import random from '../../utils/random';
+
+const LOGIN_STYPE_POPUP = 'popup';
+const LOGIN_STYPE_REDIRECT = 'redirect';
```
Comment (diegolmello) [inline app/containers/LoginServices/serviceLogin.ts:12]: Unify into TLoginStyle

File: app/containers/LoginServices/serviceLogin.ts [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+		await Services.loginOAuthOrSso({ fullName, email, identityToken });
+	} catch {
+		logEvent(events.ENTER_WITH_APPLE_F);
+	}
+};
+
+const getOAuthState = (loginStyle = LOGIN_STYPE_POPUP) => {
+	const credentialToken = random(43);
+	let obj: {
+		loginStyle: string;
```
Comment (diegolmello) [inline app/containers/LoginServices/serviceLogin.ts:144]: Apply TLoginStyle

File: app/containers/LoginServices/index.tsx [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+					collapsed={collapsed}
+					onPressButtonSeparator={onPressButtonSeparator}
+				/>
+			</>
+		);
+	}
+	return (
+		<>
+			{Object.values(services).map((service: IItemService) => (
+				<ServicesList
```
Comment (diegolmello) [inline app/containers/LoginServices/index.tsx:66]: You might need to add `key` to prevent re-renders

**Review by diegolmello — CHANGES_REQUESTED** [review]

File: app/containers/LoginServices/Service.tsx [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+		CAS_login_url: string;
+		storiesTestOnPress?: () => void;
+	}) => {
+		const { colors } = useTheme();
+
+		const { name } = service;
+		let modifiedName = name === 'meteor-developer' ? 'meteor' : name;
+		const icon = `${modifiedName}-monochromatic` as TIconsName;
+		const isSaml = service.service === 'saml';
+		let onPress: any = () => {};
```
Comment (GleidsonDaniel) [inline app/containers/LoginServices/Service.tsx:33]: Will this value be changed?
Why is it not recommended to use let inside components, as you have no reference to which value that is, which causes some re-rendering.
Same buttonText

**Review by GleidsonDaniel — CHANGES_REQUESTED** [review]

**Review by GleidsonDaniel — APPROVED** [review]

**Pull request merged**